sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs

```diff
@@ -2829,6 +2949,36 @@ public abstract partial class VectorSearchAlgorithmConfiguration
         public static bool operator !=(Azure.Search.Documents.Indexes.Models.VectorSearchAlgorithmMetric left, Azure.Search.Documents.Indexes.Models.VectorSearchAlgorithmMetric right) { throw null; }
         public override string ToString() { throw null; }
     }
+    public partial class VectorSearchProfile
+    {
+        public VectorSearchProfile(string name, string algorithm) { }
+        public string Algorithm { get { throw null; } set { } }
+        public string Name { get { throw null; } set { } }
+        public string Vectorizer { get { throw null; } set { } }
+    }
+    public abstract partial class VectorSearchVectorizer
+    {
+        protected VectorSearchVectorizer(string name) { }
+        public string Name { get { throw null; } set { } }
```

Comment:
(Applies to other models' parameters as well) Why is this settable? Required parameters as properties should be read-only per guidelines, I thought. /cc @KrzysztofCwalina @tg-msft

Reply:
This is the generated code, and I believe the generator generates constructor parameters for required properties, along with setters.

You can find an example in the guidelines [here](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-model-types) where `key` and `value` are defined as required parameters and have setters.

Reply:
But that's only if the addressable (referencable) resources can have it's name changed. Seems changing the name of a vectorizer either references or creates an entirely new resources. The guidelines you referenced I think are also wrong that `Key` shouldn't be read-write but read-only. It's not a one-size-fits-all, but depends on the context. Maybe we don't have enough knowledge in TypeSpec and/or swagger, though.

Let me start a thread offline since I'm not getting eyes from those I flagged.

Reply:
After creating an index, I can simply update the vectorizer name (as shown in the example below) and reuse existing properties. Wouldn't it be valid to have a setter for this use case?
```c#
SearchIndex searchIndex = new(indexName)
{
    Fields =
    {
        new SimpleField("HotelId", SearchFieldDataType.String) { IsKey = true },
        new SearchField("DescriptionVector", SearchFieldDataType.Collection(SearchFieldDataType.Single))
        {
            IsSearchable = true,
            VectorSearchDimensions = 1536,
            VectorSearchProfile = "my-vector-profile"
        }
    },
    VectorSearch = new()
    {
        Profiles =
        {
            new VectorSearchProfile("my-vector-profile", "my-hsnw-vector-config")
            {
                Vectorizer = "openai"
            },
            new VectorSearchProfile("my-vector-profile", "my-knn-config")
            {
                Vectorizer = "newopenai"
            }
        },
        Algorithms =
        {
            new HnswVectorSearchAlgorithmConfiguration("my-hsnw-vector-config"),
            new ExhaustiveKnnVectorSearchAlgorithmConfiguration("my-knn-config")
        },
        Vectorizers =
        {
            new AzureOpenAIVectorizer(name: "openai")
            {
                AzureOpenAIParameters = new AzureOpenAIParameters()
                {
                    ResourceUri = "Endpoint",
                    ApiKey = "key",
                    DeploymentId = "gpt-4-32k"
                }
            }
        }
    }
};

SearchIndex createdIndex = await indexClient.CreateIndexAsync(searchIndex);

AzureOpenAIVectorizer vectorizer = createdIndex.VectorSearch.Vectorizers[0] as AzureOpenAIVectorizer;
vectorizer.Name = "newopenai";

createdIndex.VectorSearch.Vectorizers[0] = vectorizer;
await indexClient.CreateOrUpdateIndexAsync(createdIndex);
```

Reply:
Yes that would be a valid change. It would essentially be a "drop" of "openai" vectorizer in the first profile, and "add" in the second profile. (Although the profile names should be different here highlighted in the image)

Query vectorizers are able to be mutated as much as the customer would like (deleted, changed, added) because it only affects query-time behavior, where the vector index essentially acts as read-only. 

![image](https://github.com/Azure/azure-sdk-for-net/assets/32447910/d868a69b-3664-434c-9782-be38e36958b2)

Reply:
Technically, yes, it works. But from a pure C# perspective, you've taking an object representing one resource and made that same object - a reference object - reference a different resource. Imagine a more simple scenario:

```c#
record class Person(string Name);

Person p = new("Shivangi");
someDbContext.AddPerson(p);
p.Name = "Heath";
someDbContext.AddPerson(p);
```
It works, sure, but what if `p` was used elsewhere. The reference has changed to a new person. It's not logically the same `p`. But if @KrzysztofCwalina or @tg-msft are okay with it, I won't stand in the way (need to review other changes).

But I still think this should be a key property in the swagger. Were it TypeSpec, it should be decorated with a `@key("name")` decorator.

Reply:
@tg-msft in my example, I called out how they shouldn't have duplicate names, and the 1 and 2 I drew onto the screenshot were meant to be suffixes onto the names. Internally we don't allow this scenario to happen. It just happens that the logic was not in place when @ShivangiReja tested it.

Reply:
Since I'm not too familiar with the SDK work, my comments would be from the perspective of our service receiving a request

```json
"vectorSearch": {
  ...
  "vectorizers": [
    {
      "name": "openai",
      "kind": "AzureOpenAI"
      "AzureOpenAIParameters": { etc...}
    }
  ]
}
```

And then an update request

```json
"vectorSearch": {
  ...
  "vectorizers": [
    {
      "name": "different-name",
      "kind": "AzureOpenAI"
      "AzureOpenAIParameters": { etc...}
    }
  ]
}
```

This update request is valid for vectorizers.

Reply:
I've created a follow up work item to make this readonly before GA.